engine/dlib/src/dlib/profile/profile.cpp

```diff
@@ -14,52 +14,563 @@
+static ProfileProperty  g_ProfileProperties[PROFILER_MAX_NUM_PROPERTIES];
+
+// *************************************************************************************************
+// Helpers
+
+static bool IsProfileInitialized()
+{
+    return dmAtomicGet32(&g_ProfileInitialized) != 0;
+}
+
+static void CreateRegisteredProperties();
+static void PropertyInitialize();
+
+// *************************************************************************************************
+// C api
+
+// *************************************************************************************************
+// Listeners
+
+static void CreateListener(ProfileListener* profiler)
+{
+    profiler->m_Ctx = 0;
+    if (profiler->m_Create)
+    {
+        profiler->m_Ctx = profiler->m_Create();
+    }
+    profiler->m_Disabled = profiler->m_Ctx == 0;
+}
+
+static void CreateProperty(ProfileListener* profiler, ProfileProperty* prop)
+{
+    #define CREATE_TYPE(STRTYPE) if (profiler->m_CreateProperty ## STRTYPE ) profiler->m_CreateProperty ## STRTYPE (profiler->m_Ctx, prop->m_Name, prop->m_Desc, prop->m_DefaultValue.m_ ## STRTYPE, prop->m_Flags, prop->m_Idx, prop->m_ParentIdx)
+
+    switch(prop->m_Type)
+    {
+    case PROFILE_PROPERTY_TYPE_GROUP:   if (profiler->m_CreatePropertyGroup) profiler->m_CreatePropertyGroup(profiler->m_Ctx, prop->m_Name, prop->m_Desc, prop->m_Idx, prop->m_ParentIdx); break;
+    case PROFILE_PROPERTY_TYPE_U32:     CREATE_TYPE(U32); break;
+    case PROFILE_PROPERTY_TYPE_S32:     CREATE_TYPE(S32); break;
+    case PROFILE_PROPERTY_TYPE_F32:     CREATE_TYPE(F32); break;
+    case PROFILE_PROPERTY_TYPE_U64:     CREATE_TYPE(U64); break;
+    case PROFILE_PROPERTY_TYPE_S64:     CREATE_TYPE(S64); break;
+    case PROFILE_PROPERTY_TYPE_F64:     CREATE_TYPE(F64); break;
+    case PROFILE_PROPERTY_TYPE_BOOL:    CREATE_TYPE(Bool); break;
+    }
+
+    #undef CREATE_TYPE
+}
+
+static void CreateListenerProperties(ProfileListener* profiler)
+{
+    for (uint32_t i = 0; i < g_ProfilePropertyCount; ++i)
+    {
+        ProfileProperty* prop = &g_ProfileProperties[i];
+        CreateProperty(profiler, prop);
+    }
+}
+
+void ProfileRegisterProfiler(const char* name, ProfileListener* profiler)
+{
+    {
+        // Check for duplicates
+        ProfileListener* p = g_ProfileListeners;
+        while (p)
+        {
+            if (strcmp(name, p->m_Name) == 0)
+            {
+                dmLogError("Profiler '%s' already registered", name);
+                return;
+            }
+            p = p->m_Next;
+        }
+    }
+
+    DM_MUTEX_OPTIONAL_SCOPED_LOCK(g_ProfileLock);
+
+    dmSnPrintf(profiler->m_Name, sizeof(profiler->m_Name), "%s", name);
+    profiler->m_Next = g_ProfileListeners; // it's ok to add it first
+    profiler->m_Ctx = 0;
+    g_ProfileListeners = profiler;
+
+    if (IsProfileInitialized())
+    {
+        CreateListener(profiler);
+        CreateListenerProperties(profiler);
+    }
```

Comment:
If the profiler is registered after the profile is initialized, it needs to get access to all the properties.